tool/tsh/db.go

```diff
@@ -62,10 +107,123 @@ func onListDatabases(cf *CLIConf) error {
+	}
+
+	allowedUsers, deniedUsers := listAllowedDeniedDbUsers(roleSet)
+
+	allUsers := make([]string, 0)
+	allUsers = append(allUsers, allowedUsers...)
+	allUsers = append(allUsers, deniedUsers...)
+
+	usersOk := make([]string, 0)
+	usersBad := make([]string, 0)
+
+	for _, user := range allUsers {
+		err := roleSet.CheckAccess(database,
+			services.AccessMFAParams{Verified: true},
+			&services.DatabaseUserMatcher{User: user})
+
+		if err == nil {
+			usersOk = append(usersOk, user)
+		} else {
+			if user != types.Wildcard {
+				usersBad = append(usersBad, user)
+			}
+		}
+	}
+
+	if len(usersOk) == 0 {
+		return "(none allowed)"
+	}
+
+	if len(usersBad) == 0 {
+		return fmt.Sprintf("%v", usersOk)
+	}
+
+	return fmt.Sprintf("%v, except: %v", usersOk, usersBad)
+}
+
+// splitAllowDeny is a helper function for splitting the users allow/deny map (true/false) into separate, sorted slices.
+func splitAllowDeny(inputMap map[string]bool) ([]string, []string) {
+	allows := make([]string, 0)
+	denies := make([]string, 0)
+
+	for entity, allow := range inputMap {
+		if allow {
+			allows = append(allows, entity)
+		} else {
+			denies = append(denies, entity)
+		}
+	}
+
+	sort.Strings(allows)
+	sort.Strings(denies)
+
+	return allows, denies
+}
+
+// listAllowedDenied works on a given role set and generic entity function getter to return a minimal description of allowed set of names.
+// It is biased towards *allowed* entities (e.g. usernames); It is meant to describe what the user can do, rather than cannot do.
+// For that reason if the user isn't allowed to pick *any* entities, the output will be empty.
+func listAllowedDenied(set services.RoleSet, getEntities func(role types.Role, conditionType types.RoleConditionType) []string) ([]string, []string) {
+	entities := make(map[string]bool)
+
+	// get allowed and denied users. denied users overwrite allowed ones.
+	for _, role := range set {
+		for _, name := range getEntities(role, types.Allow) {
+			entities[name] = true
+		}
+	}
+
+	for _, role := range set {
+		for _, name := range getEntities(role, types.Deny) {
+			entities[name] = false
+		}
+	}
+
+	// do we have a wildcard?
+	allowAll, found := entities[types.Wildcard]
+
+	if found {
+		if allowAll {
+			// allow *: all names are allowed, except for those denied. keep explicitly allowed names so user can see them.
+			return splitAllowDeny(entities)
+		}
+		// deny *: nothing is allowed. return empty lists.
+		return []string{}, []string{}
+	}
+
+	// no wildcard in either allow or deny.
+	// return users, but only the allowed ones.
+	allowed, _ := splitAllowDeny(entities)
+	return allowed, []string{}
+}
+
+func listAllowedDeniedDbUsers(set services.RoleSet) ([]string, []string) {
+	getEntities := func(role types.Role, conditionType types.RoleConditionType) []string {
+		return role.GetDatabaseUsers(conditionType)
+	}
+
+	return listAllowedDenied(set, getEntities)
+}
+
+func listAllowedDeniedDbNames(set services.RoleSet) ([]string, []string) {
```

Comment:
Yeah, I was considering adding another column with allowed db names, in which case this would come handy. I'll probably drop it and simplify the other function too.